Update `!include` YAML function. Add unit tests. Update docs (#1179)
cloudposse/atmos: website/docs/core-concepts/stacks/yaml-functions/include.mdx

```yaml
components:
  terraform:
    app:
      vars: !include config/dev.tfvars
```
</File>

<File title="stacks/org/staging/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/staging.tfvars
```
</File>

<File title="stacks/org/prod/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/prod.tfvars
```
</File>

This pattern allows you to plug Atmos into your existing Terraform/OpenTofu root modules with minimal changes — no need to
duplicate or reformat your varfiles. You also unlock additional capabilities like [listing all your stacks](/cli/commands/list/stacks)
and [components](/cli/commands/list/components), leveraging layered configurations, [stack inheritance with imports](/core-concepts/stacks/imports),
and consistent promotion of settings across environments.

## Usage

The `!include` function can be called with either one or two parameters:

```yaml
  # Download the file and inject the content directly into the current location in the YAML
  !include <file-path>

  # Download the file, filter the content using the YQ expression,
  # and inject the result directly into the current location in the YAML

>   !include <file-path> <yq-expression>

Reviewer: Never mind.

import Intro from '@site/src/components/Intro'
import File from '@site/src/components/File'
import Note from "@site/src/components/Note"

<Intro>

>     The `!include` Atmos YAML function allows downloading local or remote files from different sources,
>     and assigning the file contents or individual values to the sections in Atmos stack manifests.

Reviewer: ```suggestion
    The `!include` function in Atmos YAML lets you load files—either local or remote—and insert their contents or specific values directly into sections of your stack manifests.
```

</Intro>

> ## Description

Reviewer: ```suggestion
```

- Allows you to use [YQ](https://mikefarah.gitbook.io/yq) expressions to query and filter the content of the files to retrieve individual values.

- Automatically detects the format of the files regardless of the file extensions. It supports files in JSON, YAML
and [HCL](https://github.com/hashicorp/hcl) (`tfvars`) formats, and automatically converts them into correct
YAML structures (simple and complex types like maps and lists are supported).
All other files are returned unchanged, allowing you, for example, to include text and
[Markdown](https://www.markdownguide.org/) files as strings in Atmos manifests.

Atmos resolves the `!include` functions during the initial loading of YAML files from the local filesystem or remote sources,
injecting the contents of the referenced files directly into the current location. This applies to all Atmos configurations,
including CLI and stack configs. Included data (JSON, YAML, HCL, or text) is parsed and converted into the appropriate
type (`string`, `boolean`, `map`, `list`, etc.) before being assigned to its place in the configuration.

> ## Supported File Formats

Reviewer: ```suggestion
## Supported File Formats

With `!include` it's possible to import multiple different file formats into Atmos configurations. These formats include:
- HCL (e.g. `.tfvars` files)
- YAML
- JSON
- Text

Note that Atmos does not look at file extensions and determines the type of file from magic strings. That means that even if you have a `.txt` file with JSON, it will be decoded as JSON. 

## Supported Sources
```

> The `!include` function supports the following local file formats:

Reviewer: ```suggestion
The `!include` function supports the following local file sources:
```

The `!include` directive enables modular, reusable configuration patterns across all Atmos YAML files — both CLI and stack configurations.
This has several important implications:

- **Modularization & Reuse**  
  `!include` supports a [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) approach by allowing shared configuration fragments to be stored in local or remote
   files and reused across multiple stacks or components. You can use it in place of YAML anchors, which don't work across files.

- **Preprocessing for Inheritance**  
  Includes are resolved *before* Atmos processes stacks and components, enabling powerful [inheritance](/core-concepts/stacks/inheritance/) and deep-mergin
  behaviors using fully expanded configuration data.

- **Adopting Existing Terraform/OpenTofu Root Modules as Atmos Components**  
  If you're already managing your root modules using `.tfvars` files — for example, separate files for `dev`, `staging`,
  and `prod` — you can reference them directly in Atmos using `!include`. This makes it easy to adopt Atmos without
  having to translate all variables into Atmos stacks.

### Example: Referencing Different TFVAR Files Per Environment

> For example, instead of rewriting existing `.tfvars` varfiles into inline YAML, Atmos lets you bring them in as-is.

Reviewer: ```suggestion
For example, instead of rewriting existing `.tfvars` varfiles into inline YAML, Atmos lets you bring them in as-is into your [Stacks](/core-concepts/stacks/).
```

You can continue managing your root modules as you always have, while gaining Atmos features like stack inheritance,
environment promotion, and deep merging.

Let's say you already have environment-specific `.tfvars` files like:

- `config/dev.tfvars`
- `config/staging.tfvars`
- `config/prod.tfvars`

> You can keep using these files by referencing them in your stack configurations:

Reviewer: ```suggestion
You can keep using these files in Atmos by referencing them in your stack configurations:
```

<File title="stacks/org/dev/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/dev.tfvars
```
</File>

<File title="stacks/org/staging/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/staging.tfvars
```
</File>

<File title="stacks/org/prod/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/prod.tfvars
```
</File>

This pattern allows you to plug Atmos into your existing Terraform/OpenTofu root modules with minimal changes — no need to

> duplicate or reformat your varfiles. You also unlock additional capabilities like listing all your stacks, leveraging

Reviewer: ```suggestion
duplicate or reformat your varfiles. You also unlock additional capabilities like [listing all your stacks](/cli/commands/list/stacks) and [components](/cli/commands/list/components), leveraging
```

> layered configurations, stack inheritance, and consistent promotion of settings across environments.

Reviewer: ```suggestion
layered configurations, [stack inheritance with imports](/core-concepts/stacks/imports), and consistent promotion of settings across environments.
```

    and assigning the file contents or individual values to the sections in Atmos stack manifests.
</Intro>

> Atmos resolves the `!include` functions during the initial phase of YAML stack manifest processing.
> At this stage, it evaluates and substitutes the contents of included files, converting them into the appropriate data
> types (e.g., string, boolean, map, or list) before assigning them to the corresponding sections or attributes in the
> Atmos stack configurations.

Reviewer: ```suggestion
Atmos resolves `!include` functions during the initial loading of YAML files, injecting the contents of the referenced files directly into the current location. This applies to all Atmos configurations, including CLI and stack configs. Included data (JSON, YAML, HCL, or text) is parsed and converted into the appropriate type (`string`, `boolean`, `map`, `list`, etc.) before being assigned to its place in the configuration.
```

Key implications of this behavior:

- The `!include` functions facilitate stack configuration reuse by enabling modularization of common configurations.
  These configurations can reside in separate local or remote files and be referenced in multiple stack definitions,
  promoting a [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) approach.

- The functions preprocess YAML files before Atmos resolves stacks and components.
  This enables configuration [Inheritance](/core-concepts/stacks/inheritance/) and deep-merging of stack parameters
  after the `!include` directives are expanded.

- When transitioning from plain Terraform/OpenTofu to Atmos, `!include` allows seamless reuse of existing variable
  files (`varfiles`) by referencing them in the `vars` sections of Atmos components:

  ```yaml
    my-component:
      vars: !include my-component.tfvars

>   ```

Reviewer: ~~~suggestion
## Key Benefits of `!include`

The `!include` directive enables modular, reusable configuration patterns across all Atmos YAML files—both CLI and stack configurations. This has several important implications:

- **Modularization & Reuse**  
  `!include` supports a [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) approach by allowing shared configuration fragments to be stored in local or remote files and reused across multiple stacks or components. You can use it in place of YAML anchors, which don't work across files.

- **Preprocessing for Inheritance**  
  Includes are resolved *before* Atmos processes stacks and components, enabling powerful [inheritance](/core-concepts/stacks/inheritance/) and deep-merging behaviors using fully expanded configuration data.

- **Adopting Existing Terraform/OpenTofu Root Modules as Atmos Components**  
  If you're already managing your root modules using `.tfvars` files—for example, separate files for `dev`, `staging`, and `prod`—you can reference them directly in Atmos using `!include`. This makes it easy to adopt Atmos without having to translate all variables into Atmos stacks.

### Example: Referencing Different TFVAR Files Per Environment

For example, instead of rewriting existing `.tfvars` varfiles into inline YAML, Atmos lets you bring them in as-is. You can continue managing your root modules as you always have, while gaining Atmos features like stack inheritance, environment promotion, and deep merges.

Let's say you already have environment-specific `.tfvars` files like:

- `config/dev.tfvars`
- `config/staging.tfvars`
- `config/prod.tfvars`

You can keep using these files by referencing them in your stack configurations:

<File title="stacks/org/dev/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/dev.tfvars
```
</File>

<File title="stacks/org/staging/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/staging.tfvars
```
</File>

<File title="stacks/org/prod/app.yaml">
```yaml
components:
  terraform:
    app:
      vars: !include config/prod.tfvars
```
</File>

This pattern allows you to plug Atmos into your existing Terraform root modules with minimal changes—no need to duplicate or reformat your varfiles. You also unlock additional capabilities like listing all your stacks, leveraging layered configurations, stack inheritance, and consistent promotion of settings across environments.
~~~

    and assigning the file contents or individual values to the sections in Atmos stack manifests.
</Intro>

> Atmos resolves `!include` functions during the initial loading of YAML files from the local filesystem or remote sources, injecting the contents of the referenced files directly into the current location. This applies to all Atmos configurations, including CLI and stack configs. Included data (JSON, YAML, HCL, or text) is parsed and converted into the appropriate type (`string`, `boolean`, `map`, `list`, etc.) before being assigned to its place in the configuration.

Reviewer: Note, that we do support `!env` in Atmos CLI configuration.

```
</File>

<File title="stack.yaml">
```yaml

> import: !include config/import.yaml

Reviewer: Somewhere we need to explain the search paths to find `config/import.yaml`. What is it relative to? e.g. the component, the current stack config file, the base path, etc?

Reviewer: Let's move the part about how it loads files into a chapter titled `## Import Search Paths`. 

Let's rename "Supported File Protocols" to "Supported File Formats" and make it only about the types of files it handles. 

Also, I believe it doesn't respect file extensions - we should list that. It guesses the type of format based on the file's contents. Also, can we support a raw format? IMO, if the file ends in `.txt` this is how it should be handled.

Lastly, we should have a chapter on `## Remote Includes` and how that works and mention token injection.

    <dd>(Optional) [YQ](https://mikefarah.gitbook.io/yq) expression to retrieve individual values from the file</dd>
</dl>

## Remote Includes

Atmos uses [`go-getter`](https://github.com/hashicorp/go-getter)
(used by [Terraform](https://www.terraform.io/) for downloading modules) to download remote files from different sources.

### Downloading from OCI Registries

The following example shows how to include a file from a public AWS ECR registry:

  ```yaml
    components:
      terraform:
        my-component:

>           vars: !include oci://public.ecr.aws/cloudposse/components/terraform/stable/aws/my-component.tfvars

Reviewer: We shouldn't use this example any more because the repo was deleted.

Reviewer: I think the most common would be from a repo, not from OCI. We should provide the most helpful/practical example first.
